File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-0.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        memo = {}
        high = 0
        def bfs(x, y):
            if (x, y) in memo:
                return memo[(x, y)]
            cur = matrix[x][y]
            ans = 1
            if x + 1 < len(matrix) and matrix[x + 1][y] > cur:
                ans = max(ans, 1 + bfs(x + 1, y))
            if x - 1 >= 0 and matrix[x - 1][y] > cur:
                ans = max(ans, 1 + bfs(x - 1, y))
            if y + 1 < len(matrix[0]) and matrix[x][y + 1] > cur:
                ans = max(ans, 1 + bfs(x, y + 1))
            if y - 1 >= 0 and matrix[x][y - 1] > cur:
                ans = max(ans, 1 + bfs(x, y - 1))
            memo[(x, y)] = ans
            return ans
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                high = max(high, bfs(i, j))
        return high

File: Data Structures & Algorithms/count-number-of-islands/submission-0.cpp
class Solution {
public:
    void bfs(int x, int y, vector<vector<char>>& grid){
        if (x < 0 || x >= grid.size() || y < 0 || y >= grid[0].size() || grid[x][y] != '1'){
            return;
        }
        grid[x][y] = '0';
        bfs(x + 1,y,grid);
        bfs(x - 1,y,grid);
        bfs(x,y + 1,grid);
        bfs(x,y - 1,grid);
    }
    int numIslands(vector<vector<char>>& grid) {
        int cnt = 0;
        for (int i = 0;i < grid.size();i++){
            for(int j = 0;j < grid[0].size();j++){
                if (grid[i][j] == '1'){
                    bfs(i,j,grid);
                    cnt++;
                }
            }
        }
        return cnt;
    }
};
